lib/pbench/server/api/resources/query_apis/datasets_publish.py

```diff
@@ -0,0 +1,253 @@
+import json
+from http import HTTPStatus
+from logging import Logger
+
+from flask_restful import abort
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources.query_apis import (
+    API_OPERATION,
+    CONTEXT,
+    ElasticBase,
+    JSON,
+    PostprocessError,
+    Schema,
+    Parameter,
+    ParamType,
+)
+from pbench.server.database.models.tracker import Dataset, Metadata
+from pbench.server.database.models.users import User
+
+
+class DatasetsPublish(ElasticBase):
+    """
+    Change the "access" authorization of a Pbench dataset
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("controller", ParamType.STRING, required=True),
+                Parameter("name", ParamType.STRING, required=True),
+                Parameter("access", ParamType.ACCESS, required=True),
+            ),
+            role=API_OPERATION.UPDATE,
+        )
+
+    def preprocess(self, client_json: JSON) -> CONTEXT:
+        """
+        Query the Dataset associated with this name, and determine whether the
+        authenticated user has UPDATE access to this dataset. (Currently, this
+        means the authenticated user is the owner of the dataset, or has ADMIN
+        role.)
+
+        If the user has authorization to update the dataset, return the dataset
+        object as CONTEXT so that the postprocess operation can mark it as
+        published.
+
+        Args:
+            json_data: JSON dictionary of type-normalized key-value pairs
+                controller: the controller that generated the dataset
+                name: name of the dataset to publish
+                access: The desired access level of the dataset (currently either
+                    "private" or "public")
+
+        Returns:
+            CONTEXT referring to the dataset object if the operation should
+            continue, or None
+        """
+        dataset = Dataset.attach(
```

Comment:
The `attach` name always seems a little confusing, its actually fetching the dataset :)

Reply:
Ha; when I started out, the idea was that I was either *creating* a new dataset "tracker" or *attaching* to an existing tracker. The word seemed meaningful at the time, and nobody commented on that during review so I didn't ever stop to re-think it.

As I've been working on other things using the `Dataset` it's started to really annoy me that I named the module `tracker.py` when it really should be `dataset.py`. In fact I almost changed it but decided that was too much of a digression from the point of this PR and would bring in other changes that complicate the review. I may still do that separately, however, just to clean things up; and that might be a good time to reconsider the method names since it'll touch the same files.